### PR TITLE
[SYCL] Fix hier_par test after pulldown from llvm

### PR DESCRIPTION
**Pull request by againull**
Signed-off-by: Artur Gainullin <artur.gainullin@intel.com>

### CODE REVIEW

File: clang/test/CodeGenSYCL/hier_par.cpp [hunk 1]
```diff
@@ -32,12 +31,11 @@ void foo() {
 
     cgh.parallel_for_work_group<class hpar_hw>(
         range<1>(1), range<1>(1), [=](group<1> g) {
-// CHECK: @[[SHADOW:[a-zA-Z0-9]+]] = internal unnamed_addr addrspace(3) global %[[GROUP_CLASS:"[^"]+"]] undef, align [[ALIGN:[0-9]+]]
-// CHECK: define {{.*}} spir_func void @{{"[^"]+"}}({{[^,]+}}, %[[GROUP_CLASS]]* byval(%[[GROUP_CLASS]]) align {{[0-9]+}} %[[GROUP_OBJ:[A-Za-z_0-9]+]]) {{.*}}!work_group_scope{{.*}} {
-// CHECK-NOT: {{^[ \t]*define}}
-// CHECK: %[[TMP:[A-Za-z_0-9]+]] = bitcast %[[GROUP_CLASS]] addrspace(3)* @[[SHADOW]] to i8 addrspace(3)*
-// CHECK:  %[[OBJ:[A-Za-z_0-9]+]] = bitcast %[[GROUP_CLASS]]* %[[GROUP_OBJ]] to i8*
-// CHECK:  call void @llvm.memcpy.p3i8.p0i8.i64(i8 addrspace(3)* align [[ALIGN]] %[[TMP]], {{[^,]+}} %[[OBJ]], {{[^)]+}})
+          // CHECK: @[[SHADOW:[a-zA-Z0-9]+]] = internal unnamed_addr addrspace(3) global %[[GROUP_CLASS:"[^"]+"]] undef, align [[ALIGN:[0-9]+]]
```
Comment (bader) [inline clang/test/CodeGenSYCL/hier_par.cpp]: Is LLVM transformation always uses `addrspace(3)` for "local" memory? I think it's supposed to be target depended.
If so, I think we should explicitly specify the target device in the RUN command to make sure that "local" memory is represented as `addrspace(3)`.

Why do we care about `align` attribute?
```suggestion
          // CHECK: @[[SHADOW:[a-zA-Z0-9]+]] = internal unnamed_addr addrspace(3) global %[[GROUP_CLASS:"[^"]+"]] undef
```

File: clang/test/CodeGenSYCL/hier_par.cpp [hunk 2]
```diff
@@ -32,12 +31,11 @@ void foo() {
...
+          // CHECK: define {{.*}} spir_func void @{{"[^"]+"}}({{[^,]+}}, %[[GROUP_CLASS]]* byval(%[[GROUP_CLASS]]) align {{[0-9]+}} %[[GROUP_OBJ:[A-Za-z_0-9]+]]) {{.*}}!work_group_scope{{.*}} {
+          // CHECK-NOT: {{^[ \t]*define}}
+          // CHECK:  %[[OBJ:[A-Za-z_0-9]+]] = bitcast %[[GROUP_CLASS]]* %[[GROUP_OBJ]] to i8*
+          // CHECK:  call void @llvm.memcpy.p3i8.p0i8.i64(i8 addrspace(3)* align [[ALIGN]] getelementptr inbounds (%[[GROUP_CLASS]], %[[GROUP_CLASS]] addrspace(3)* @[[SHADOW]], i32 0, i32 0), {{[^,]+}} %[[OBJ]], {{[^)]+}})
```
Comment (bader) [inline clang/test/CodeGenSYCL/hier_par.cpp]: What are last two checks for?

File: clang/test/CodeGenSYCL/hier_par.cpp (hunk 1 above)
Comment (againull) [reply clang/test/CodeGenSYCL/hier_par.cpp]: Good concern, thx, I think I can create different check labels for different targets to address this.

File: clang/test/CodeGenSYCL/hier_par.cpp (hunk 2 above)
Comment (againull) [reply clang/test/CodeGenSYCL/hier_par.cpp]: The last two checks are needed to ensure that group object which is passed as byval to the 'parallel for work group' is copied to 'shadow' variable which is a global variable in local address space.

Comment (bader) [reply clang/test/CodeGenSYCL/hier_par.cpp]: I've shared with you my private experiments with enabling standard LLVM transformation passes before translation to SPIR-V format. IIRC, these copies from private to local memory are optimized away. The same might happen in the back-end compilers. 
I suggest checking if it's the case. If so, the optimized code is not valid SYCL/OpenCL program.

File: clang/test/CodeGenSYCL/hier_par.cpp (hunk 1 above)
Comment (bader) [reply clang/test/CodeGenSYCL/hier_par.cpp]: I'm fine if you add a TODO and we address this later.

**Comment by bader** [comment]
Replaced by https://github.com/intel/llvm/pull/1306.

**Pull request closed without merging**